drivers/netdev/ieee802154.c

```diff
@@ -172,6 +174,33 @@ int netdev_ieee802154_get(netdev_ieee802154_t *dev, netopt_t opt, void *value,
             res = sizeof(l2filter_t **);
             break;
 #endif
+        case NETOPT_PDU_SIZE:
```

Comment:
Doesn't this need introduction?

Reply:
It already exists, but it's only used by `semtech-loramac` (and is used for something different there, so a different name would likely be needed)

Reply:
Given L2 destination length and L2 flags, we consult the L2 device of the maximum payload it can transmit.
PDU should be the entire frame. I think what we are asking for is SDU.  https://de.wikipedia.org/wiki/Service_Data_Unit

`NETOPT_MAX_SDU_SIZE` ?

Reply:
> Given L2 destination length and L2 flags, we consult the L2 device of the maximum payload it can transmit.
PDU should be the entire frame. I think what we are asking for is SDU. https://de.wikipedia.org/wiki/Service_Data_Unit
> 
> NETOPT_MAX_SDU_SIZE ?

I think it makes sense to ask for the (Service) Data Unit though the interface, as it heavily depends on the MAC state (e.g slot length, required Information Elements, etc).
As proposed in the other comment, I would try to unify such logic into `flags` rather than passing address length. Then you don't need those packing functions for the context.

Reply:
Instead of flags, another idea would be to pass the `gnrc_netif_hdr` and let the link-layer convert it to IEEE 802.15.4 header. If this does not violate some IPC or packet snip rule, this should be most flexible.
I think it is fine because the call is synchronous and the netif header must not be modified.

```C
        union {
            iolist_t in;
            uint16_t out;
        } value;
        value.in.iol_base = pkt->data; /* gnrc_netif_hdr_t */
        value.in.iol_len = pkt->size;
        if (gnrc_netapi_get(netif->pid, NETOPT_MAX_SDU_SIZE, 0, &value.in, sizeof(value)) > 0) {
            fbuf->best_frag_size = value.out; /* potential new member in gnrc_sixlowpan_frag_fb_t for best frabment size */
            assert(netif->sixlo.max_frag_size <= fbuf->best_frag_size);
        }
```

Reply:
> Instead of flags, another idea would be to pass the `gnrc_netif_hdr` and let the link-layer convert it to IEEE 802.15.4 header. If this does not violate some IPC or packet snip rule, this should be most flexible. I think it is fine because the call is synchronous and the netif header must not be modified.
> 
> ```c
>         union {
>             iolist_t in;
>             uint16_t out;
>         } value;
>         value.in.iol_base = pkt->data; /* gnrc_netif_hdr_t */
>         value.in.iol_len = pkt->size;
>         if (gnrc_netapi_get(netif->pid, NETOPT_MAX_SDU_SIZE, 0, &value.in, sizeof(value)) > 0) {
>             fbuf->best_frag_size = value.out; /* potential new member in gnrc_sixlowpan_frag_fb_t for best frabment size */
>             assert(netif->sixlo.max_frag_size <= fbuf->best_frag_size);
>         }
> ```

From a NETOPT API perspective, this assumes NETOPT_MAX_SDU_SIZE receives a `gnrc_netif_hdr_t` argument, which is not available in network stacks such as LWIP (which also implement generic `netif_***` functions). In that case I would stick to the proposed `netopt_pdu_size_ctx_unpack` function, because it would still work with other network stacks.

Is there something else we need to indicate from above `netif` besides "will use L2 dst address", in order to calculate the SDU? The source address will be set by the MAC layer anyway and (future) options will be stored close to the MAC too. Why do we need more than an L2 flag for that?

Reply:
For the normal 802.15.4 header, I would assume that destination address length, and destination PAN are sufficient.
For us destination PAN is always source PAN and thus not really needed.
To determine the size of the security header, I would need the actual destination address to know which key identifier mode is used. I know we do only implicit key mode, in which case it is always 5 bytes. However I think passing the netif header is future prove.

I thought `NETOPT_MAX_SDU_SIZE` could be implemented in `gnrc_netif_ops_t ieee802154_ops::get()`. Or is this a submac violation? So LWIP would neither know about it nor need it, right?

Reply:
I have uploaded a [branch](https://github.com/fabian18/RIOT/tree/best-6lo-frag-size) which is trying to solve it that way, just to show you what I mean. If you think it is a good alternative I could open an alternative PR. If it is not a good approach we can also stick to this.

Reply:
I like it, especially the `gnrc_netif_hdr_to_ieee802154()` function.

Feels a bit wasteful to call it twice, but that's certainly the cleanest way to do it.

Want to go ahead and PR it?